### PR TITLE
Ensure the integrity of the partial state from remote

### LINKED ISSUE #1074: Verify the integrity of PartialState
Garbage state changes should not go to the db. 

https://0chain.slack.com/archives/C01V27E2Z9N/p1645211459050089

### PR DESCRIPTION
**Pull request by peterlimg**
## Changes
- Add `StateChangesCount` filed to `Block` struct and assign the value on blocks generation.
- Validate the partial state in `ComputeProperties()` method to ensure that all nodes are connected and could be reached from root. Also check if duplicate nodes are detected, which is for preventing malicious nodes fake it with duplicate nodes to meet the `StateChangesCount` requirement.

## Fixes
- #1074 

## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- gosdk: https://github.com/0chain/gosdk/pull/422
- blobber: https://github.com/0chain/blobber/pull/609
- zboxcli: https://github.com/0chain/zboxcli/pull/205
- zwalletcli: https://github.com/0chain/zwalletcli/pull/105
- blobber-stake: https://github.com/0chain/0cloud/pull/15

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix/add-state-changes-count-block  |
| zwalletcli  | fix/add-state-changes-count-block  |
| 0chain  | fix/garbage-partial-state  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by ma2b0043** [comment]
@peterlimg @dabasov when can we expect to close this PR?

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix/add-state-changes-count-block  |
| zwalletcli  | fix/add-state-changes-count-block  |
| 0chain  | fix/garbage-partial-state  |
| blobber  | fix/state-changes-count-block  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix/add-state-changes-count-block  |
| zwalletcli  | fix/add-state-changes-count-block  |
| 0chain  | fix/garbage-partial-state  |
| blobber  | fix/state-changes-count-block  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/chaincore/state/partial_state.go [hunk 1]
```diff
@@ -97,35 +96,63 @@ func SetupPartialState(store datastore.Store) {
...
+	return nil
 }
 
-//Validate - implement interface
-func (ps *PartialState) Validate(ctx context.Context) error {
-	return ps.mndb.Validate(ps.root)
+// Validate does nothing but to meet the Entity interface, the ComputeProperties
+// have done the validation, and usually this function is called after the
+// ComputeProperties
+func (ps *PartialState) Validate(_ context.Context) error {
```
Comment (dabasov) [inline code/go/0chain.net/chaincore/state/partial_state.go:154]: maybe we can move validation here and call it in ComputeProperties()?

**Review by dabasov — DISMISSED** [review]
lgtm

File: code/go/0chain.net/chaincore/state/partial_state.go (hunk 1 above)
Comment (peterlimg) [reply code/go/0chain.net/chaincore/state/partial_state.go:154]: This function will be called after the `ComputeProperties()` in the `PutEntityHandler`, so if we call move validation code to it we will call it twice then, one for when the `ComputeProperties()` is called, one for calling the `Validation()` explicitly in the `PutEntityHandler`. 

Even though I think we might not need the `PutEntityHandler` as the `v1/chain/put` might not being used (not sure, need further checking to confirm), if so, we can remove the `Validation()` interface, and do validation in `ComputeProperties()` for all entities, but that will introduce a lot of changes, so we'd better do it  in another PR.

Comment (dabasov) [reply code/go/0chain.net/chaincore/state/partial_state.go:154]: got it

**Review by dabasov — APPROVED** [review]

**Pull request merged**